ion/reader.go

```diff
@@ -0,0 +1,388 @@
+package ion
+
+import (
+	"bufio"
+	"bytes"
+	"io"
+	"math"
+	"math/big"
+	"strings"
+	"time"
+)
+
+// A Reader reads a stream of Ion values.
+//
+// The Reader has a logical position within the stream of values, influencing the
+// values returnedd from its methods. Initially, the Reader is positioned before the
+// first value in the stream. A call to Next advances the Reader to the first value
+// in the stream, with subsequent calls advancing to subsequent values. When a call to
+// Next moves the Reader to the position after the final value in the stream, it returns
+// false, making it easy to loop through the values in a stream.
+//
+// 	var r Reader
+// 	for r.Next() {
+// 		// ...
+// 	}
+//
+// Next also returns false in case of error. This can be distinguished from a legitimate
+// end-of-stream by calling Err after exiting the loop.
+//
+// When positioned on an Ion value, the type of the value can be retrieved by calling
+// Type. If it has an associated field name (inside a struct) or annotations, they can
+// be read by calling FieldName and Annotations respectively.
+//
+// For atomic values, an appropriate XxxValue method can be called to read the value.
+// For lists, sexps, and structs, you should instead call StepIn to move the Reader in
+// to the contained sequence of values. The Reader will initially be positioned before
+// the first value in the container. Calling Next without calling StepIn will skip over
+// the composite value and return the next value in the outer value stream.
+//
+// At any point while reading through a composite value, including when Next returns false
+// to indicate the end of the contained values, you may call StepOut to move back to the
+// outer sequence of values. The Reader will be positioned at the end of the composite value,
+// such that a call to Next will move to the immediately-following value (if any).
+//
+// 	r := NewTextReaderStr("[foo, bar] [baz]")
+// 	for r.Next() {
+// 		if err := r.StepIn(); err != nil {
+// 			return err
+// 		}
+// 		for r.Next() {
+// 			fmt.Println(r.StringValue())
+// 		}
+// 		if err := r.StepOut(); err != nil {
+// 			return err
+// 		}
+// 	}
+// 	if err := r.Err(); err != nil {
+// 		return err
+// 	}
+//
+type Reader interface {
+
+	// SymbolTable returns the current symbol table, or nil if there isn't one.
+	// Text Readers do not, generally speaking, have an associated symbol table.
+	// Binary Readers do.
+	SymbolTable() SymbolTable
+
+	// Next advances the Reader to the next position in the current value stream.
+	// It returns true if this is the position of an Ion value, and false if it
+	// is not. On error, it returns false and sets Err.
+	Next() bool
```

Comment:
I was following the pattern of [sql.Rows](https://golang.org/pkg/database/sql/#Rows.Next), and personally find that style to be friendlier than (what I imagine to be) the alternative?

```go
for r.Next() {
   switch r.Type() { ... }
}

if r.Err() {
  return r.Err()
}
```

vs

```go
for {
  t := r.Next()
  if t == ion.ErrorType {
    return r.Err()
  }
  if t == ion.EndOfSequenceType {
    break
  }
  switch t { ... }
}
```

(Unless I'm misunderstanding what you're saying, which is entirely possible. :))

Reply:
I was thinking more like:
```go
for {
    switch r.Next() {
        case ion.ErrorType:
            return r.Err()
        case ion.EndOfSequenceType:
            return nil
        case ...
    }
}
```

Reply:
Which if you think about it could be simplified to

```go
for {
    switch r.Next() {
        case ion.ErrorType, ion.EndOfSequenceType:
            return r.Err()
        case ...
    }
}
```

Reply:
Ah. That's better than my attempt, although I still prefer things the sql.Rows way. Not my problem any more so feel free to change it if you disagree. :)

Reply:
I made the suggestion because it is similar to `text.Scanner` and I was thinking that the Reader is kind of like an ion.Scanner.  I don't think that me saying my suggestion was more idiomatic was fair.  They are both idiomatic approaches.

Reply:
Created #28

Reply:
I assume [Type](https://github.com/amzn/ion-go/blob/master/ion/type.go) should be mapping to Ion Types.
In case of ```EndOfSequenceType``` or ```ErrorType``` could we return ```NoType``` ?
When I check ion-dotnet, I see we have similar logic for ```MoveNext()``` in [BinaryReader](https://github.com/amzn/ion-dotnet/blob/master/Amazon.IonDotnet/Internals/Binary/UserBinaryReader.cs#L36) and [TextReader](https://github.com/amzn/ion-dotnet/blob/master/Amazon.IonDotnet/Internals/Text/RawTextReader.cs#L130).